processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.topn;
+
+import com.google.common.collect.Maps;
+import io.druid.query.aggregation.Aggregator;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.segment.Capabilities;
+import io.druid.segment.Cursor;
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.column.Column;
+
+import java.util.Map;
+
+public class TimeExtractionTopNAlgorithm extends BaseTopNAlgorithm<int[], Map<String, Aggregator[]>, TopNParams>
+{
+  public static final int[] EMPTY_INTS = new int[]{};
+  private final TopNQuery query;
+
+  public TimeExtractionTopNAlgorithm(Capabilities capabilities, TopNQuery query)
+  {
+    super(capabilities);
+    this.query = query;
+  }
+
+
+  @Override
+  public TopNParams makeInitParams(DimensionSelector dimSelector, Cursor cursor)
+  {
+    return new TopNParams(
+        dimSelector,
+        cursor,
+        Integer.MAX_VALUE, // artificially set cardinality high to force a single pass
```

Comment:
TopN is currently deeply tied to DimensionSelector, so this would require a major rewrite of topN, which I would rather leave to when we decide to support arbitrary columns as dimensions.